test/allDocsGenerator.js

```diff
@@ -0,0 +1,140 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* global describe it beforeEach */
+'use strict';
+
+const asyncGenerator = require('../includes/allDocsGenerator.js');
+const assert = require('assert');
+const request = require('../includes/request.js');
+const fs = require('fs');
+const nock = require('nock');
+const { Writable } = require('node:stream');
+const { pipeline } = require('node:stream/promises');
+
+function allDocsGen(dbUrl, opts) {
+  const db = request.client(dbUrl, opts);
+  // Disable compression to make body assertions easier
+  db.service.setEnableGzipCompression(false);
+  return asyncGenerator(db, opts);
+}
+
+describe('#unit async all docs generator', function() {
+  const dbUrl = 'http://localhost:5984/animaldb';
+
+  beforeEach('Reset nocks', function() {
+    nock.cleanAll();
+  });
+
+  it('should return expected items to writer', async function() {
+    let actualTotal = 0;
+    // Query string keys are stringified by Nano
+    const badgerKey = 'badger\0';
+    const kookaburraKey = 'kookaburra\0';
+    const snipeKey = 'snipe\0';
+    let iter = 1;
+
+    const couch = nock(dbUrl)
+      // batch 1
+      .post('/_all_docs', { limit: 3, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_1.json', { 'Content-Type': 'application/json' })
+      // batch 2
+      .post('/_all_docs', { limit: 3, start_key: badgerKey, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_2.json', { 'Content-Type': 'application/json' })
+      // batch 3
+      .post('/_all_docs', { limit: 3, start_key: kookaburraKey, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_3.json', { 'Content-Type': 'application/json' })
+      // batch 4
+      .post('/_all_docs', { limit: 3, start_key: snipeKey, include_docs: true })
+      .replyWithFile(200, './test/fixtures/animaldb_all_docs_4.json', { 'Content-Type': 'application/json' });
+
+    await pipeline(
+      allDocsGen(dbUrl, { parallelism: 1, bufferSize: 3 }),
+      new Writable({
+        objectMode: true,
+        write: (chunk, encoding, callback) => {
+          for (const [key, actualDoc] of Object.entries(chunk.docs)) {
+            const expectedDocs = JSON.parse(fs.readFileSync(`./test/fixtures/animaldb_all_docs_${iter}.json`, 'utf8'));
+            assert.deepStrictEqual(actualDoc, expectedDocs.rows[key].doc);
+          }
+          assert.ok(Number.isInteger(chunk.batch));
+          assert.ok(Number.isInteger(chunk.total));
+          if (chunk.batch === 3) {
+            assert.strictEqual(chunk.docs.length, 2); // smaller last batch
+          } else {
+            assert.strictEqual(chunk.docs.length, 3);
+          }
+          actualTotal = chunk.total;
```

Comment:
If we drop total then this should become:
```suggestion
          actualTotal += chunk.docs.length;
```